### PR TITLE
Run flutter tests through mini test engine when run directly (flutter run -t test_file)

### LINKED ISSUE #24650: `flutter run test/widget_test.dart` no longer works
(I discussed this with @jonahwilliams already, but I'm filing this to make sure we're tracking it.)

`flutter run test/widget_test.dart` is broken after the recent test_api changes.  Attempting it will hit various runtime exceptions such as:

> NoSuchMethodError: The method 'test' was called on null.
> NoSuchMethodError: The method 'setUp' was called on null.
> NoSuchMethodError: The method 'tearDown' was called on null.

[We document that people can use `flutter run` on their widget tests](https://flutter.io/docs/testing) to help debug widget tests, so this should be expected to work.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/24650

If Declarer is not defined, we need to set up our own mini-engine to execute flutter tests and log messages. Instead of adding a dep on test_core, we duplicate a subset of the functionality needed for flutter run test.  To intercept all of the calls, we hide test, group, setUp, setUpAll, tearDown, tearDownAll and re-export the same methods.

This code does not run when tests are executed via `flutter test` since the declarer is already defined.

This also adds an integration test that confirms that the output of flutter run -t works as expected.

### CODE REVIEW

**Review by yjbanov — APPROVED** [review]

File: dev/devicelab/bin/tasks/flutter_run_test.dart [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+const Pattern skippedMessageMatch = '+1 -1: example skipped';
+const Pattern finishedMessageMatch = '+1 ~1 -1: Some tests failed.';
+
+Future<void> main() async {
+  deviceOperatingSystem = DeviceOperatingSystem.android;
+  await task(createFlutterRunTask());
+}
+
+TaskFunction createFlutterRunTask() {
+  return () async {
```
Comment (yjbanov) [inline dev/devicelab/bin/tasks/flutter_run_test.dart]: Do you need an inner closure? Why not put the logic directly inside the top-level function then tear it off?

Comment (yjbanov) [reply dev/devicelab/bin/tasks/flutter_run_test.dart]: i.e. `await task(flutterRunTask);`

**Comment by yjbanov** [comment]
Is there a reason why this needs to run in the devicelab? It does not seem to need any devices, so could run in Cirrus? LGTM otherwise.

**Comment by jonahwilliams** [comment]
This specifically verifies that flutter run -t test_file works, running a test via flutter test will go through the old paths since the declarer will be defined.

**Comment by yjbanov** [comment]
lgtm

File: dev/devicelab/bin/tasks/flutter_run_test.dart (hunk 1 above)
Comment (jonahwilliams) [reply dev/devicelab/bin/tasks/flutter_run_test.dart]: A result of some over-eager copy pasting, fixed

**Comment by jonahwilliams** [comment]
cc @Hixie

File: dev/devicelab/bin/tasks/flutter_run_test.dart [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  flutterDirectory.path, 'dev', 'automated_tests', 'flutter_run_test', 'flutter_run_test.dart',
+));
+const Pattern passedMessageMatch = '+0: example passed';
+const Pattern failedMessageMatch = '+1: example failed [E]';
+const Pattern skippedMessageMatch = '+1 -1: example skipped';
+const Pattern finishedMessageMatch = '+1 ~1 -1: Some tests failed.';
+
+Future<void> main() async {
+  deviceOperatingSystem = DeviceOperatingSystem.android;
+  await createFlutterRunTask();
```
Comment (yjbanov) [inline dev/devicelab/bin/tasks/flutter_run_test.dart]: You still need `await task(createFlutterRunTask)`.

File: packages/flutter_test/lib/src/test_compat.dart [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+import 'package:test_api/src/backend/group.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/group_entry.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/test.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/suite.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/live_test.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/suite_platform.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/runtime.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/message.dart'; // ignore: implementation_imports
+import 'package:test_api/src/backend/invoker.dart';  // ignore: implementation_imports
+import 'package:test_api/src/backend/state.dart'; // ignore: implementation_imports
```
Comment (Hixie) [inline packages/flutter_test/lib/src/test_compat.dart]: well this is going to cost us a lot in technical debt...

Comment (Hixie) [reply packages/flutter_test/lib/src/test_compat.dart]: this seems like a flaw in the test_api package. How are we supposed to do this?

File: packages/flutter_test/lib/src/test_compat.dart [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+}
+
+Future<void> _runLiveTest(Suite suiteConfig, LiveTest liveTest, _Reporter reporter, {bool countSuccess = true}) async {
+  reporter._onTestStarted(liveTest);
+  // Schedule a microtask to ensure that [onTestStarted] fires before the
+  // first [LiveTest.onStateChange] event.
+  await Future<void>.microtask(liveTest.run);
+  // Once the test finishes, use [new Future] to do a coarse-grained event
+  // loop pump to avoid starving non-microtask events.
+  await Future<void>(() {});
```
Comment (Hixie) [inline packages/flutter_test/lib/src/test_compat.dart]: `await null` should do the same thing.

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 5]
```diff
@@ -19,12 +19,19 @@ import 'controller.dart';
...
 /// Keep users from needing multiple imports to test semantics.
 export 'package:flutter/rendering.dart' show SemanticsHandle;
 
 export 'package:test_api/test_api.dart' hide
+  test,
+  group,
+  setUpAll,
+  tearDownAll,
+  setUp,
+  tearDown,
```
Comment (Hixie) [inline packages/flutter_test/lib/src/widget_tester.dart]: add comments explaining why (as for the existing hidden members)

**Comment by Hixie** [comment]
As we keep reimplementing more and more of the test package, I see less and less reason to bother with it at all...

File: packages/flutter_test/lib/src/test_compat.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_test/lib/src/test_compat.dart]: These APIs are stable but haven't been formally made public yet. I've opened https://github.com/dart-lang/test/issues/962 to track, but most of these could be moved to lib.

**Pull request merged**